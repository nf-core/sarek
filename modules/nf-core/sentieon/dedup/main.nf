process SENTIEON_DEDUP {
    tag "$meta.id"
    label 'process_medium'
    label 'sentieon'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://community-cr-prod.seqera.io/docker/registry/v2/blobs/sha256/a6/a64461f38d76bebea8e21441079e76e663e1168b0c59dafee6ee58440ad8c8ac/data' :
        'community.wave.seqera.io/library/sentieon:202308.03--59589f002351c221' }"

    input:
    tuple val(meta), path(bam), path(bai)
    tuple val(meta2), path(fasta)
    tuple val(meta3), path(fasta_fai)

    output:
    tuple val(meta), path("*.cram")               , emit: cram, optional: true
    tuple val(meta), path("*.crai")               , emit: crai, optional: true
    tuple val(meta), path("*.bam")                , emit: bam , optional: true
    tuple val(meta), path("*.bai")                , emit: bai
    tuple val(meta), path("*.score")              , emit: score
    tuple val(meta), path("*.metrics")            , emit: metrics
    tuple val(meta), path("*.metrics.multiqc.tsv"), emit: metrics_multiqc_tsv
    path "versions.yml"                           , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def args2 = task.ext.args2 ?: ''
    def args3 = task.ext.args3 ?: ''
    def args4 = task.ext.args4 ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}.cram"
    def metrics = task.ext.metrics ?: "${prefix}.metrics"
    def input_list = bam.collect{"-i $it"}.join(' ')
    def prefix_basename = prefix.substring(0, prefix.lastIndexOf("."))
    def sentieonLicense = secrets.SENTIEON_LICENSE_BASE64 ?
        "export SENTIEON_LICENSE=\$(mktemp);echo -e \"${secrets.SENTIEON_LICENSE_BASE64}\" | base64 -d > \$SENTIEON_LICENSE; " :
        ""
    """
    $sentieonLicense

    sentieon driver $args -t $task.cpus $input_list -r ${fasta} --algo LocusCollector $args2 --fun score_info ${prefix_basename}.score
    sentieon driver $args3 -t $task.cpus $input_list -r ${fasta} --algo Dedup $args4 --score_info ${prefix_basename}.score --metrics ${metrics} ${prefix}

    # This following tsv-file is produced in order to get a proper tsv-file with Dedup-metrics for importing in MultiQC as "custom content".
    # It should be removed once MultiQC has a module for displaying Dedup-metrics.
    head -3 ${metrics} > ${metrics}.multiqc.tsv

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        sentieon: \$(echo \$(sentieon driver --version 2>&1) | sed -e "s/sentieon-genomics-//g")
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}.cram"
    def metrics = task.ext.metrics ?: "${prefix}.metrics"
    def prefix_basename = prefix.substring(0, prefix.lastIndexOf("."))

    """
    touch "${prefix}"
    touch "${prefix}.crai"
    touch "${prefix}.bai"
    touch "${metrics}"
    touch "${metrics}.multiqc.tsv"
    touch "${prefix_basename}.score"

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        sentieon: \$(echo \$(sentieon driver --version 2>&1) | sed -e "s/sentieon-genomics-//g")
    END_VERSIONS
    """
}
