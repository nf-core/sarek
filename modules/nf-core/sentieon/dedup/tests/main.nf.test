nextflow_process {

    name "Test Process SENTIEON_DEDUP"
    tag "modules_nfcore"
    tag "modules"
    tag "sentieon"
    tag "dedup"
    tag "sentieon/dedup"

    script "../main.nf"
    process "SENTIEON_DEDUP"

    test("Test marking duplicates") {
        config "./nextflow.config"

        when {
            process {
                """
                input[0] = [
                    [ id:'test' ], // meta map
                    [ file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true) ],
                    [ file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam.bai', checkIfExists: true) ]
                ]
                input[1] = [[id: 'test'],file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)]
                input[2] = [[id: 'test'],file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta.fai', checkIfExists: true)]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("Test removing duplicates") {
        config "./nextflow_rmdup.config"

        when {
            process {
                """
                input[0] = [
                    [ id:'test' ], // meta map
                    [ file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true) ],
                    [ file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam.bai', checkIfExists: true) ]
                ]
                input[1] = [[id: 'test'],file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)]
                input[2] = [[id: 'test'],file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta.fai', checkIfExists: true)]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("Test stub") {
        config "./nextflow.config"
        options "-stub"
        when {
            process {
                """
                input[0] = [
                    [ id:'test' ], // meta map
                    [ file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.single_end.sorted.bam', checkIfExists: true) ],
                    [ file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.single_end.sorted.bam.bai', checkIfExists: true) ]
                ]
                input[1] = [[id: 'test'],file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)]
                input[2] = [[id: 'test'],file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta.fai', checkIfExists: true)]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }
}
