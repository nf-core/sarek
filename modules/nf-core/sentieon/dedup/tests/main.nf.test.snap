{
    "Test marking duplicates": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.cram:md5,e46e97256846338e1cff32d862105491"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test.cram.crai:md5,4b7b2152b33c5334f9477cc3650f8c91"
                    ]
                ],
                "2": [
                    
                ],
                "3": [
                    [
                        {
                            "id": "test"
                        },
                        "test.cram.bai:md5,889503338dc569b24e44e5e3aec815ea"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test"
                        },
                        "test.score:md5,835f05ecc5d3ef5d4e31ba7f831d9a8b"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test"
                        },
                        "test.cram.metrics:md5,208f7c5fa2f489cfaaffbce116fed0bc"
                    ]
                ],
                "6": [
                    [
                        {
                            "id": "test"
                        },
                        "test.cram.metrics.multiqc.tsv:md5,208f7c5fa2f489cfaaffbce116fed0bc"
                    ]
                ],
                "7": [
                    "versions.yml:md5,763463853476be96846b6da5aecfacf4"
                ],
                "bai": [
                    [
                        {
                            "id": "test"
                        },
                        "test.cram.bai:md5,889503338dc569b24e44e5e3aec815ea"
                    ]
                ],
                "bam": [
                    
                ],
                "crai": [
                    [
                        {
                            "id": "test"
                        },
                        "test.cram.crai:md5,4b7b2152b33c5334f9477cc3650f8c91"
                    ]
                ],
                "cram": [
                    [
                        {
                            "id": "test"
                        },
                        "test.cram:md5,e46e97256846338e1cff32d862105491"
                    ]
                ],
                "metrics": [
                    [
                        {
                            "id": "test"
                        },
                        "test.cram.metrics:md5,208f7c5fa2f489cfaaffbce116fed0bc"
                    ]
                ],
                "metrics_multiqc_tsv": [
                    [
                        {
                            "id": "test"
                        },
                        "test.cram.metrics.multiqc.tsv:md5,208f7c5fa2f489cfaaffbce116fed0bc"
                    ]
                ],
                "score": [
                    [
                        {
                            "id": "test"
                        },
                        "test.score:md5,835f05ecc5d3ef5d4e31ba7f831d9a8b"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,763463853476be96846b6da5aecfacf4"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-02T10:28:10.570152622"
    },
    "Test removing duplicates": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.cram:md5,8075d3e7c66d36fdbb81270eefc996d4"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test.cram.crai:md5,c617398ead281c1339d78d5df0d606e9"
                    ]
                ],
                "2": [
                    
                ],
                "3": [
                    [
                        {
                            "id": "test"
                        },
                        "test.cram.bai:md5,a1ea729eca4732ca3a5dee946a70fbc8"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test"
                        },
                        "test.score:md5,835f05ecc5d3ef5d4e31ba7f831d9a8b"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test"
                        },
                        "test.cram.metrics:md5,2a41239de0275a8321f4658286d97d65"
                    ]
                ],
                "6": [
                    [
                        {
                            "id": "test"
                        },
                        "test.cram.metrics.multiqc.tsv:md5,2a41239de0275a8321f4658286d97d65"
                    ]
                ],
                "7": [
                    "versions.yml:md5,763463853476be96846b6da5aecfacf4"
                ],
                "bai": [
                    [
                        {
                            "id": "test"
                        },
                        "test.cram.bai:md5,a1ea729eca4732ca3a5dee946a70fbc8"
                    ]
                ],
                "bam": [
                    
                ],
                "crai": [
                    [
                        {
                            "id": "test"
                        },
                        "test.cram.crai:md5,c617398ead281c1339d78d5df0d606e9"
                    ]
                ],
                "cram": [
                    [
                        {
                            "id": "test"
                        },
                        "test.cram:md5,8075d3e7c66d36fdbb81270eefc996d4"
                    ]
                ],
                "metrics": [
                    [
                        {
                            "id": "test"
                        },
                        "test.cram.metrics:md5,2a41239de0275a8321f4658286d97d65"
                    ]
                ],
                "metrics_multiqc_tsv": [
                    [
                        {
                            "id": "test"
                        },
                        "test.cram.metrics.multiqc.tsv:md5,2a41239de0275a8321f4658286d97d65"
                    ]
                ],
                "score": [
                    [
                        {
                            "id": "test"
                        },
                        "test.score:md5,835f05ecc5d3ef5d4e31ba7f831d9a8b"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,763463853476be96846b6da5aecfacf4"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-02T10:28:19.377946074"
    },
    "Test stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.cram:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test.cram.crai:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    
                ],
                "3": [
                    [
                        {
                            "id": "test"
                        },
                        "test.cram.bai:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test"
                        },
                        "test.score:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test"
                        },
                        "test.cram.metrics:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "6": [
                    [
                        {
                            "id": "test"
                        },
                        "test.cram.metrics.multiqc.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "7": [
                    "versions.yml:md5,763463853476be96846b6da5aecfacf4"
                ],
                "bai": [
                    [
                        {
                            "id": "test"
                        },
                        "test.cram.bai:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "bam": [
                    
                ],
                "crai": [
                    [
                        {
                            "id": "test"
                        },
                        "test.cram.crai:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "cram": [
                    [
                        {
                            "id": "test"
                        },
                        "test.cram:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "metrics": [
                    [
                        {
                            "id": "test"
                        },
                        "test.cram.metrics:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "metrics_multiqc_tsv": [
                    [
                        {
                            "id": "test"
                        },
                        "test.cram.metrics.multiqc.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "score": [
                    [
                        {
                            "id": "test"
                        },
                        "test.score:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,763463853476be96846b6da5aecfacf4"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-02T10:28:28.792696026"
    }
}